### PR TITLE
fix: resolve post-variants per iteration in editor canvas

### LINKED ISSUE #604: Bug: post-variants render on every iteration in the canvas and start empty when added
## Expected Behavior

In the Query Loop block's editor canvas:

1. **Per-post resolution.** Each post iteration should render only the variant that *matches* its index (or the base post-template content when no variant matches). A `position:first` variant should only appear on post #1; other posts should keep showing the base template.
2. **Inherited inner blocks on add.** Creating a new variant via the "Post Variants" panel should seed the variant's inner blocks with a deep clone of the post-template's current base children (title, cover, etc.) — not start empty.
3. **Selecting a variant.** Clicking a variant in the "Post Variants" panel should auto-jump the editable iteration to the first post that variant matches, so the canvas is WYSIWYG.

## Current Behavior

In the canvas (see screenshot):

1. The `position:first` variant's content (title "testing" + cover block) is rendered on **every** post tile in the loop, alongside the base template. Each tile also shows the `position:first` chip/summary. Result: every post looks like the variant, which is the opposite of the intended per-post override.
2. Adding a variant via the "Post Variants" panel inserts an empty `artisanpack/post-variant` block. The user has to re-add title, image, etc. from scratch — even though the user almost always wants to start from the base template and tweak.
3. Selecting a variant doesn't change which iteration is editable.

Net effect: variants are unusable in practice. The compiled page render is fine (`QueryInliner` does the right thing), but the editor preview misleads authors into thinking every post will look identical, and the per-variant authoring loop is full of busywork.

## Steps to Reproduce

1. In the dev app, open the visual editor on a page that contains a Query Loop block with a Post Template that has > 1 child block (e.g. `post-title`, then an empty `cover` block).
2. Select the Query Loop block → Inspector → "Post Variants" panel → "Add position variant".
3. Open the new variant and add some content (e.g. a `post-title` with the text "testing", plus a `cover` block).
4. Observe the canvas: every post in the loop shows the `position:first` chip, the title "testing", and the cover block — not just the first.
5. Also note: when the variant was added in step 2, its body was empty; the user had to re-author the title + cover from scratch instead of starting from the post-template's existing children.

**Expected:** Only post #1 renders the variant; posts #2..N render the base template.
**Actual:** Every post renders the union of `(base template + every variant's children)`, and every variant starts empty when added.

## Environment

- **Operating System:** macOS (Darwin 25.5.0) — but root cause is editor JS; reproducible on any host.
- **Browser:** Any modern browser (bug is in `@wordpress/block-editor` integration code).
- **PHP Version:** 8.4.3 (not directly relevant — server-side `QueryInliner` is correct).
- **Project Version:** `main` of `visual-editor` package (commit `1294b84`).

### Root cause

Two distinct bugs, same feature area:

**(1) Editor canvas doesn't resolve variants per-iteration.**

`resources/js/visual-editor/editor/query-preview-iterations.tsx` reads the post-template's entire `innerBlocks` (which includes every `artisanpack/post-variant` child) and passes that flat list to both the editable `<InnerBlocks />` and the per-iteration `<PreviewIteration>` ghosts. There's no equivalent of the server-side `QueryInliner::extractVariants()` + `VariantResolver` here, so every iteration renders the base blocks *and* every variant's wrapper / content.

The matching/resolution primitives already exist in `resources/js/visual-editor/editor/variant-matcher.ts` — `sortVariants`, `compileStaticMap`, `resolveVariant` — and `PostVariantsPanel` already builds the descriptor list and writes the static map. The preview pipeline just doesn't consume them.

**(2) New variants are created empty.**

`resources/js/visual-editor/blocks/query/post-variants-panel.tsx:186-195` calls:

```ts
const block = createBlock( 'artisanpack/post-variant', {
    matcher: defaultMatcherFor( kind ),
    priority: 10,
} );
```

No third argument → no inner blocks. The user almost always wants to start from the base template and customize.

### Fix approach

**Fix (1) — per-iteration variant resolution in the canvas:**

In `query-preview-iterations.tsx`:

- After selecting `innerBlocks`, partition them into `baseChildren` (non-variant) and `variantBlocks` (`name === 'artisanpack/post-variant'`), mirroring `QueryInliner::extractVariants()`.
- Build a `VariantDescriptor[]` from `variantBlocks` (using each block's attributes for `matcher`, `priority`, `label`, and its document-order index for `order`).
- Read the compiled static map off the post-template's `_compiledVariantMap` attribute (it's already maintained by `PostVariantsPanel`).
- For each post in `visiblePosts`, call `resolveVariant(index, total, postMeta, descriptors, staticMap)` to find the winning variant's `order` (or `null` for "use base").
- Render that iteration's `<PreviewIteration>` with the resolved children: `variantBlocks[order].innerBlocks` if matched, else `baseChildren`.
- For the editable iteration: drop the variant blocks from the `<InnerBlocks />` template so they no longer render inline on every post. The variant's own editable surface is the variant block in the list view / when selected.

**Fix (2) — seed new variants from the base template:**

In `post-variants-panel.tsx:handleAdd`:

- Look up the post-template block (already known: `postTemplateClientId`) and read its current non-variant inner blocks via `select('core/block-editor').getBlock(postTemplateClientId).innerBlocks`.
- Filter out any `artisanpack/post-variant` children, then deep-clone each via `cloneBlock` from `@wordpress/blocks` (regenerates clientIds so the clones are independent instances).
- Pass the cloned array as the third arg to `createBlock( 'artisanpack/post-variant', attrs, clonedBaseChildren )`.
- After insert, dispatch `selectBlock(newVariantClientId)` so the inspector immediately focuses the new variant's controls.

**Fix (3) — auto-jump editable iteration to first matching post:**

In `query-preview-iterations.tsx`:

- When the selected block is an `artisanpack/post-variant` (or one of its descendants), compute the first iteration index that resolves to that variant's `order` and set `activeId` to that post's id.
- Subscribe to the selection store (`useSelect` against `core/block-editor`'s `getSelectedBlockClientId` / `getBlockParentsByBlockName`) so this stays reactive.
- Don't override the user's explicit clicks on other ghosts (i.e. only auto-jump when the selection changes, not on every render).

### Acceptance / regression tests

- Add a Pest/Vitest test under `resources/js/visual-editor/editor/__tests__/query-preview-iterations.test.tsx` that mounts the component with: 5 preview posts, a `position:first` variant with title="testing" + cover children, and asserts only iteration #1 renders the variant tree (the other four render the base template).
- Add a test for a `pattern:odd` variant — iterations 1, 3, 5 should render the variant; 2, 4 should render base.
- Add a test in `resources/js/visual-editor/blocks/query/__tests__/post-variants-panel.test.tsx` that asserts `handleAdd` produces a variant block whose `innerBlocks` deep-equal (modulo clientId) the post-template's current non-variant children.
- Add a test that the auto-jump behavior fires when a variant is selected and that user-clicked ghost activations aren't overridden by later selection changes that still resolve to the same iteration.

### Affected files

- `packages/visual-editor/resources/js/visual-editor/editor/query-preview-iterations.tsx` — main fix site for per-iteration resolution + auto-jump.
- `packages/visual-editor/resources/js/visual-editor/blocks/query/post-variants-panel.tsx:186-195` — fix `handleAdd` to seed inner blocks from the base template.
- `packages/visual-editor/resources/js/visual-editor/editor/variant-matcher.ts` — no changes; already provides `resolveVariant` / `compileStaticMap`. Document its consumption here.
- `packages/visual-editor/resources/js/visual-editor/blocks/post-template/edit.tsx` — likely fine as-is since it just forwards to `QueryPreviewIterations`, but verify the `defaultTemplate` doesn't end up duplicated in variants after the fix.
- `packages/visual-editor/resources/js/visual-editor/blocks/post-variant/edit.tsx` — verify the editable variant's `<InnerBlocks />` body still renders correctly when selected; no other changes expected.

### Reference: server-side parity

`packages/visual-editor/src/Resources/QueryInliner.php:1180-1201` already splits base vs variants correctly via `extractVariants()`, and `VariantResolver` handles per-post resolution. The fix is to bring the editor preview into parity with that pipeline.

### Severity

**High** — the feature exists and was shipped, but it's effectively unusable in the editor: authors can't preview what they're building, and adding a variant is busywork. No data loss (server render is correct), but the authoring experience is broken enough that no one would reach for variants today.

## Acceptance Criteria

- [ ] In the editor canvas, a `position:first` variant only renders on iteration #1; remaining iterations render the base post-template children.
- [ ] `pattern:odd` / `pattern:even` / `every-nth:<step>` variants resolve correctly per iteration in the canvas.
- [ ] `meta:has-featured-image` resolves correctly using the preview post payload.
- [ ] Adding a variant via the Post Variants panel seeds its inner blocks with a deep clone of the post-template's current non-variant children (independent clientIds).
- [ ] Selecting a variant in the Post Variants panel (or in list view) auto-jumps the editable iteration to the first post that variant matches.
- [ ] User-clicked ghost activations are not overridden by selection-driven auto-jumps that still resolve to the same iteration.
- [ ] Unit tests added for per-iteration variant resolution (position + pattern + meta cases) and for the seed-on-add behavior.
- [ ] Manual smoke: open a query loop with > 1 post, add a `position:first` variant, edit its title, verify only post #1 changes in the canvas and that the saved page matches the canvas preview.

## Implementation Tasks

- [ ] `query-preview-iterations.tsx`: partition `innerBlocks` into base + variants, build descriptor list, read `_compiledVariantMap` off the post-template, resolve per-iteration via `resolveVariant`, and render the resolved children (variant or base) in each ghost.
- [ ] `query-preview-iterations.tsx`: drop variant blocks from the editable iteration's `<InnerBlocks />` so they don't render inline on every post.
- [ ] `query-preview-iterations.tsx`: subscribe to `core/block-editor` selection and auto-jump `activeId` when a variant is selected (only when the new selection wouldn't already match the active post).
- [ ] `post-variants-panel.tsx:handleAdd`: read the post-template's non-variant inner blocks, deep-clone them via `cloneBlock`, and pass them as the third arg to `createBlock( 'artisanpack/post-variant', attrs, clones )`. Then dispatch `selectBlock(newVariantClientId)`.
- [ ] Add Vitest specs under `editor/__tests__/query-preview-iterations.test.tsx` covering position, pattern, and meta resolution paths.
- [ ] Add Vitest spec under `blocks/query/__tests__/post-variants-panel.test.tsx` covering the new seed-from-base behavior.

### PR DESCRIPTION
**Pull request by ViewFromTheBox**
## Description

Fixes two distinct bugs in the Query Loop **Post Variants** feature that together made variants unusable in the editor: every variant rendered on every post tile in the canvas, and adding a variant started from a blank slate. The compiled page render was already correct via `QueryInliner` — only the editor preview pipeline was broken.

**Closes:** #604

## Type of Change

- [x] Bug fix (fixes an issue)

## Related Issue

**Issue:** #604

## Motivation and Context

The editor preview misled authors into thinking every post would look identical to the variant, and creating a variant via the "Post Variants" panel inserted an empty `artisanpack/post-variant` block — forcing the user to re-author title/cover/etc. from scratch. Net effect: variants existed but no one would reach for them.

## Changes Made

- `resources/js/visual-editor/editor/query-preview-iterations.tsx`
  - Partition the post-template's `innerBlocks` into base children + `artisanpack/post-variant` children, build a `VariantDescriptor[]`, read `_compiledVariantMap`, and call `resolveVariant` per iteration.
  - Ghost iterations now render the resolved set via `useBlockPreview` (variant `innerBlocks` or base) — so a `position:first` variant only previews on iteration #1.
  - Editable iteration carries `data-resolved-variant-order`; an injected `<style>` collapses non-matching children (base blocks when a variant is resolved, other variants either way) so the active post shows only its resolved content.
  - Subscribe to the block-editor selection store; when a `artisanpack/post-variant` (or descendant) is selected, auto-jump `activeId` to the first iteration that matches that variant. A ref guard prevents overriding the user's most recent manual ghost activation.
- `resources/js/visual-editor/blocks/query/post-variants-panel.tsx`
  - `handleAdd` now deep-clones the post-template's current non-variant inner blocks via `cloneBlock` (independent clientIds, recursive), seeds them into the new variant, and `selectBlock`s the new variant so its inspector opens immediately.
- New vitest specs:
  - `editor/__tests__/query-preview-iterations.test.tsx` — adds 5 cases (`position:first`, `pattern:odd`, `meta:has-featured-image`, instance static map, auto-jump on variant selection).
  - `blocks/query/__tests__/post-variants-panel.test.tsx` — 4 cases covering seed-from-base (including nested `innerBlocks`), excluding existing variants from the seed, and auto-selecting the new variant; also empty-base behavior.

## How Has This Been Tested?

**Testing Environment:**
- Operating System: macOS (Darwin 25.5.0)
- Browser: Safari (latest)
- PHP Version: 8.4.3
- Project Version: visual-editor `release/1.2`

**Tests Performed:**
1. `npm test -- --run resources/js/visual-editor/editor/__tests__/query-preview-iterations.test.tsx resources/js/visual-editor/blocks/query/__tests__/post-variants-panel.test.tsx` — 21 passing (17 + 4).
2. Manual canvas test in the dev app — added a `pattern:odd` variant with a custom paragraph + post-title; only odd iterations (1, 3, 5) render the variant; even iterations (2, 4) render the base template; switching the editable iteration via ghost click flips the collapse correctly; adding a new variant pre-seeds with the base template's children; the new variant is auto-selected with its inspector open.
3. Hard-refresh in the canvas iframe — no Rules-of-Hooks regressions, no crashes.

## Accessibility Tests Run

- [x] Keyboard navigation tested — ghost iterations still respond to Enter/Space activation.
- [ ] Screen reader tested — no announce-level changes; `<style>` injection is presentational.
- [x] Color contrast verified — no color changes.
- [x] ARIA labels checked — unchanged from prior implementation.

## Tests Added

- [x] Unit tests added/updated
- [x] All tests passing

**Test details:**
- `query-preview-iterations.test.tsx`: position / pattern / meta / instance-map resolution + auto-jump.
- `post-variants-panel.test.tsx`: seed-from-base deep-clone (incl. nested), exclusion of existing variants, select-after-insert, empty-base edge case.

## Documentation

- [x] Inline code documentation added (file-level docblock updated to call out #604 semantics)

## Pre-Submission Checklist

- [x] Followed contributing guidelines
- [x] Checked for other open PRs for same update
- [x] Code passes all tests (pre-existing site-editor-app localStorage failures unrelated to this change)
- [x] Self-review completed
- [x] Comments added for complex code

## New Features
- Post variants now clone and inherit the non-variant blocks from the parent post template when created, including nested block structure.
- Variant matching now supports position, pattern, and featured-image conditions, and applies the resolved variant per post iteration.
- Editing focus now auto-jumps to the first post matching the currently selected variant, while previews correctly show non-editable “ghost” iterations.

## Tests
- Added and expanded automated coverage for variant seeding and per-iteration variant resolution behavior.

### CODE REVIEW

File: resources/js/visual-editor/blocks/query/post-variants-panel.tsx [hunk 1]
```diff
@@ -187,11 +188,32 @@ export default function PostVariantsPanel( {
...
+        const block = createBlock(
+            'artisanpack/post-variant',
+            {
+                matcher: defaultMatcherFor( kind ),
+                priority: 10,
+            },
+            baseChildren
+        );
         insertBlock( block, variantRows.length, postTemplateClientId );
+        // Focus the new variant so its inspector controls open
```
[the remark this answers is not in the document]
Comment (ViewFromTheBox) [reply resources/js/visual-editor/blocks/query/post-variants-panel.tsx:228]: The fix has been applied in d741d6d — the file now uses `insertIndex` (computed via `lastVariantIndex + 1` or `templateChildren.length`), not `variantRows.length`. Local `cr review` returns 0 findings; this comment is a stale re-anchor on the patched line.

File: resources/js/visual-editor/editor/query-preview-iterations.tsx [hunk 2]
```diff
@@ -155,8 +296,88 @@ export function QueryPreviewIterations(
...
+            rules.push(
+                'li[data-query-iteration="editable"][data-resolved-variant-order="' +
+                    String( order ) +
+                    '"] > .block-editor-inner-blocks > .block-editor-block-list__layout > :not([data-block="' +
+                    cid +
+                    '"]){display:none!important;}'
+            );
+        } );
+        return rules.join( '' );
+    }, [ variantBlocks ] );
```
[the remark this answers is not in the document]
Comment (ViewFromTheBox) [reply resources/js/visual-editor/editor/query-preview-iterations.tsx]: Already addressed in d741d6d via `data-query-preview-root="{clientId}"` on the outer wrapper + a `scopedRootSelector` prefix on every collapse rule. The comment is anchored to the now-removed unscoped selectors (line=None in the API). Local `cr review` returns 0 findings.
